code/parse/sexp_container.cpp

```diff
@@ -787,3 +787,261 @@ const char *sexp_container_CTEXT(int node)
+		if (map_data.find(possible_key) == map_data.end()) {
+			return SEXP_FALSE;
+		}
+
+		node = CDR(node);
+	}
+
+	return SEXP_TRUE;
+}
+
+/**
+* Check if a SEXP map container has a key whose data matches a specific string.
+* If a third argument (string variable) is provided, store the data's associated key, if a match is found.
+* Returns -1 if the element is not in the container
+*/
+int sexp_map_has_data_item(int node)
+{
+	const char *container_name = CTEXT(node);
+	const auto *p_container = get_sexp_container(container_name);
+
+	if (!p_container) {
+		Warning(LOCATION, "Map-has-data-item called on nonexistent container %s.", container_name);
+		return SEXP_FALSE;
+	}
+
+	const auto &container = *p_container;
+
+	if (!container.is_map()) {
+		Warning(LOCATION, "Map-has-data-item called on non-map container %s.", container_name);
+		return SEXP_FALSE;
+	}
+
+	node = CDR(node);
+	Assertion(node != -1, "Map-has-data-item wasn't given data to look for. Please report!");
+
+	const SCP_string possible_data = CTEXT(node);
+
+	for (const auto &kv_pair : container.map_data) {
+		if (possible_data == kv_pair.second) {
+			// check for optional variable to store the key
+			node = CDR(node);
+			if (node != -1) {
+				int var_index = -1;
+				if (Sexp_nodes[node].type & SEXP_FLAG_VARIABLE) {
+					var_index = sexp_get_variable_index(node);
+				} else {
+					// perhaps the node text is the variable name as data
+					var_index = get_index_sexp_variable_name(Sexp_nodes[node].text);
+				}
+
+				if (var_index >= 0) {
+					const int var_type = Sexp_variables[var_index].type;
+					if (((var_type & SEXP_VARIABLE_STRING) && any(container.type & ContainerType::STRING_KEYS)) ||
+						((var_type & SEXP_VARIABLE_NUMBER) && any(container.type & ContainerType::NUMBER_KEYS))) {
+						// assign key to variable
+						sexp_modify_variable(kv_pair.first.c_str(), var_index);
+					} else {
+						Warning(LOCATION,
+							"Map-has-data-item given optional variable %s whose type doesn't match keys of map "
+							"container %s",
+							Sexp_variables[var_index].variable_name,
+							container.container_name.c_str());
+					}
+				} else {
+					Warning(LOCATION, "Map-has-data-item given invalid optional variable %s", Sexp_nodes[node].text);
+				}
+			}
+
+			return SEXP_TRUE;
+		}
+	}
+
+	return SEXP_FALSE;
+}
+
+int sexp_container_eval_status_sexp(int op_num, int node)
+{
+	switch (op_num) {
+		case OP_IS_CONTAINER_EMPTY:
+			return sexp_is_container_empty(node);
+			break;
+
+		case OP_GET_CONTAINER_SIZE:
+			return sexp_get_container_size(node);
+			break;
+
+		case OP_LIST_HAS_DATA:
+			return sexp_list_has_data(node);
+			break;
+
+		case OP_LIST_DATA_INDEX:
+			return sexp_list_data_index(node);
+			break;
+
+		case OP_MAP_HAS_KEY:
+			return sexp_map_has_key(node);
+			break;
+
+		case OP_MAP_HAS_DATA_ITEM:
+			return sexp_map_has_data_item(node);
+			break;
+
+		default:
+			UNREACHABLE("Unknown container status SEXP operator %d", op_num);
```

Comment:
Just needs a please report, imho.

Reply:
Good catch, I updated the `Assert()` messages but forgot to update the `UNREACHABLE()` ones.